UART_driver/Core/Src/CircularBuffer.cpp

```diff
@@ -0,0 +1,50 @@
+#include "CircularBuffer.hpp"
+#include <cstring>
+
+CircularBuffer::CircularBuffer(uint8_t* buf, uint16_t size) {
+    this->buf = buf;
+    this->size = size;
+}
+
+uint8_t CircularBuffer::peek(uint8_t& res, uint16_t dist) {
+    if(dist + readPtr >= writePtr){
+        return 0;
+    }
+    res = buf[(readPtr + dist) % size];
+    return 1;
+}
```

Comment:
Ok so two things here. Since you are returning error codes, it may be beneficial to create an enum to represent the different types of errors? Mainly a readability thing instead of returning magic numbers (1 or 0).

Another thing, is generally we like to keep to 1 return statement. I believe it is apart of some safety standard but generally what you see is, if you create the enum that im talking about above ^ you would first initialize that variable then set it during the function call and fianlly return it at the end in one return statement.

Sometimes that may lead to a bunch of nested if else so im ok with having multiple returns there but just try to avoid it if possible (like here)

Reply:
The error code thing goes for the entire code written in this PR.

Reply:
One return statement is part of Misra C iirc. Each function should only have one exit point so there's never a chance of someone missing an exit point/case.

Reply:
@Chrisc110 I'm thinking if it would be easier to just return `bool` signaling read/write success. For both read and write there is only one failure case which is reading beyond the available data or writing when full, so there's no need for complicated enums.

Reply:
So I would be ok with this just be bools are primitive data types and people wouldn’t have to be worried about including the header that has the definition.

I think this is less your problem but rather a system level thing where we define one file called like “error.h” that has all the possible errors that we can run into and is common across everything so everyone can use it.

So in conclusion, that sounds good to me!